composition-js/src/compositionOptions.ts

```diff
@@ -0,0 +1,60 @@
+import {
+  didYouMean,
+  DirectiveDefinition,
+  ERRORS,
+  executableDirectiveLocations,
+  Subgraphs,
+  suggestionList,
+  type SubtypingRule,
+} from "@apollo/federation-internals";
+import { GraphQLError } from 'graphql';
+
+export type CompositionOptions = {
+  mergeDirectives?: string[],
+  allowedFieldTypeMergingSubtypingRules?: SubtypingRule[],
+};
+
+export const validateCompositionOptions = (toMerge: Subgraphs, options?: CompositionOptions): GraphQLError[] => {
+  // for mergeDirectives, we want to validate that every directive specified starts with a '@'
+  // and exists on some subgraph. Also ensure that none of the directives are builtin or federation directives
+  const subgraphs = toMerge.values();
+  const mergeDirectives = options?.mergeDirectives ?? [];
+  const errors: GraphQLError[] = [];
+  mergeDirectives.forEach(directiveName => {
+    if (directiveName[0] !== '@') {
+      errors.push(ERRORS.INVALID_MERGE_DIRECTIVES_ARGUMENT.err({
+        message: `Directive "${directiveName}" in "mergeDirectives" argument does not begin with a "@"`,
+      }));
+    } else {
+      const directiveNameWithoutAt = directiveName.slice(1);
+
+      // for the directive specified, get the DirectiveDefinition for each subgraph it appears in
+      const subgraphDirectives = subgraphs
+        .map(sg => sg.schema.directive(directiveNameWithoutAt))
+        .filter((directive): directive is DirectiveDefinition => directive !== undefined);
+
+      if (subgraphDirectives.length === 0) {
+        // If the directive does not appear in any subgraph, throw an error. Provide a suggestion if we think it's a typo.
```

Comment:
Agreed. Some `IGNORED_MERGED_DIRECTIVES_ARGUMENT` "warning" hint could make sense, both for this one, but also the next 2 cases (built-ins and only-executable). In all 3 cases, we just want to say that the provided value is essentially not gonna do anything, but a warning is arguably probably even a bit better than a hard error.

Tbc, I'd still keep a hard error for the lacking of `@` (it's a syntax error, no need to be lenient) and the `no core stuff`, because it's meant to be supported eventually and so ignoring with a warning feels super dodgy for future upgrades.

Reply:
Unless you strongly object, I think I'll leave the hard error for the next two since it seems best to just force them to get the arguments correct.